README.md
express-mquery
====================

[![Build Status](https://travis-ci.com/lykmapipo/express-mquery.svg?branch=master)](https://travis-ci.com/lykmapipo/express-mquery)
[![Dependencies Status](https://david-dm.org/lykmapipo/express-mquery.svg)](https://david-dm.org/lykmapipo/express-mquery)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/express-mquery/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/express-mquery?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/express-mquery)](https://github.com/lykmapipo/express-mquery/blob/master/LICENSE)

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/express-mquery)](https://www.npmjs.com/package/express-mquery)

Expose [mongoose](https://github.com/Automattic/mongoose) query API through HTTP request with partial support of [json-api](http://jsonapi.org/): 
 - [sparse-fieldsets](http://jsonapi.org/format/#fetching-sparse-fieldsets)
 - [sorting](http://jsonapi.org/format/#fetching-sorting)
 - [pagination](http://jsonapi.org/format/#fetching-pagination)
 - [filtering](http://jsonapi.org/format/#fetching-filtering)
 - [includes](http://jsonapi.org/format/#fetching-includes)

*Note: Checkout the current [specification](https://github.com/lykmapipo/express-mquery/blob/master/SPECIFICATION.md) for more information.*

## Installation
```js
$ npm install --save express-mquery
```

## Usage
```js
import expess from 'express';
import mquery from 'express-mquery';

const app = express();
app.use(mquery({ limit: 10, maxLimit: 50 }));

...

app.get('/users', (request, response, next) => {
  
  // obtain request.mquery
  console.log(request.mquery);

});
```

## Structure
Once parse, `express-mquery` will extend `http request` with `mquery` field

Example:

```js
GET /invoices?fields=number,amount&filter[name]=Bob
&filter[amount][$gte]=1200&include=customer,items
&fields[customer]=name,number&fields[items]=name,price
&page[number]=1&page[size]=10&sort[number]=1&sort[amount]=-1
```

Will be parsed into:
```js
{
  filter: { name: "Bob", amount: { $gte: 1200 } },
  paginate: { limit: 10, skip: 0, page: 1 },
  populate: [
  	{ path: "customer", select: { name: 1, number: 1 } },
  	{ path: "items", select: { name: 1, price: 1 } }
  ],
  select: { number: 1, amount: 1 },
  sort: { number: 1, amount: -1 }
}
```

Where:
- `filter` : Is valid `mongoose` criteria and can be passed to `find()`
- `paginate` : Contains paging details that can be passed to `limit()`, `skip()`
- `populate` : Is valid `mongoose` populate option and can be passed to `populate()`
- `select` : Is valid `mongoose` project options and can be passed to `select()`
- `sort` : Is valid `mongoose` sort options and can be passed to `sort()`


## Querying

>When passing values as objects or arrays in URLs, they must be valid JSON

### Sort
```js
GET /customers?sort=name
GET /customers?sort=-name
GET /customers?sort={"name":1}
GET /customers?sort={"name":1, "email":-1}

or

GET /customers?sort=name
GET /customers?sort=-name
GET /customers?sort[name]=1&sort[email]=-1
```

### Page
```js
GET /customers?page=1
GET /customers?page=1&limit=10
GET /customers?page[number]=1&page[size]=10
```

### Skip
```js
GET /customers?skip=10
```

### Limit
Only overrides `maximum limit option set by the plugin` if the queried limit is lower
```js
GET /customers?limit=10
```

### Query or Filters
Supports all [mongodb operators](https://docs.mongodb.com/manual/reference/operator/query/) `($regex, $gt, $gte, $lt, $lte, $ne, etc.)`

```js
GET /customers?query={"name":"Bob"}
GET /customers?query={"name":{"$regex":"/Bo$/"}}
GET /customers?query={"age":{"$gt":12}}
GET /customers?query={"age":{"$gte":12}}

or

GET /customers?filter[name]=Bob
GET /customers?filter[name][$regex]="/Bo$/"
GET /customers?filter[age][$gt]=12
GET /customers?filter[age][$gte]=12
```

### Populate or Include
Works with create, read and update operations

```js
GET /invoices?populate=customer
GET /invoices?populate={"path":"customer"}
GET /invoices?populate=[{"path":"customer"},{"path":"products"}]

or

GET /invoices?include=customer
GET /invoices?include[customer]=name,number&includes[items]=name,price
GET /invoices?include=customer,items&fields[customer]=name,number&fields[items]=name,price
```

### Select or Fields
`_id` is always returned unless explicitely excluded

```js
GET /customers?select=name
GET /customers?select=-name
GET /customers?select={"name":1}
GET /customers?select={"name":0}

or

GET /customers?fields=name
GET /customers?fields=-name
GET /customers?fields=name,email
GET /invoices?include=customer&fields[customer]=name
```

## Testing

* Clone this repository

* Install `grunt-cli` global

```sh
$ npm install -g grunt-cli
```

* Install all development dependencies

```sh
$ npm install
```

* Then run test

```sh
$ npm test
```

## Contribute

Fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## Licence

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 


CHANGELOG.md
#### 1.3.1 (2021-08-23)

##### Chores

* **deps:**  force latest version & audit fix ([f603a40c](https://github.com/lykmapipo/express-mquery/commit/f603a40cc6aaf69e59e8a4af69f8c5ea00c5b750))

##### Documentation Changes

* **readme:**  update build status to travis.com ([21f9fda2](https://github.com/lykmapipo/express-mquery/commit/21f9fda257d178222f3d7cc1455bb7446f7a0bb3))

#### 1.3.0 (2021-08-05)

##### Chores

*  move lint config file to test folder ([d1c373fe](https://github.com/lykmapipo/express-mquery/commit/d1c373fee1bf85f318724a03ccd8be71d76015b7))
*  install deps and rollup configuration ([d51f5a72](https://github.com/lykmapipo/express-mquery/commit/d51f5a7223455a1664df60dfa7899d18b73c5f77))
*  setup editor configuration ([4760cebf](https://github.com/lykmapipo/express-mquery/commit/4760cebf82e6282b4669fc73442e23cd517a77e7))
*  delete unused files ([dbd73bd7](https://github.com/lykmapipo/express-mquery/commit/dbd73bd761f46913bd9ba3bbd1734f5267a583d1))
*  setup eslint and prettier ([8c805426](https://github.com/lykmapipo/express-mquery/commit/8c805426a601645a6353916295e468d60121fe62))
* **deps:**  bump lodash from 4.17.20 to 4.17.21 ([9ef73590](https://github.com/lykmapipo/express-mquery/commit/9ef73590304fda49e673272a8417b77c5ec8399a))

##### Documentation Changes

*  update status badge & license year ([2b0313a3](https://github.com/lykmapipo/express-mquery/commit/2b0313a39eb27307db7ff87637db0c4677c6db39))
*  add code of conduct & contributing guide ([4a969214](https://github.com/lykmapipo/express-mquery/commit/4a969214278ebad7e0586ca19602087e02094eef))
* **readme:**
  *  update usage ([de74dcc5](https://github.com/lykmapipo/express-mquery/commit/de74dcc557b14ae6342022712e76526a758d7237))
  *  add link to specifications ([2607699f](https://github.com/lykmapipo/express-mquery/commit/2607699f598c4e4b7879978ae29b43a65750b6a0))

##### Other Changes

* //github.com/rijkerd/express-mquery into rijkerd-feature/migrate-to-es6 ([f2350133](https://github.com/lykmapipo/express-mquery/commit/f2350133320a10a3db79601d5096ef91c117bee8))
*  install deps ([4a822850](https://github.com/lykmapipo/express-mquery/commit/4a822850313407f7bb5f6f9e1c13161b059a02d1))
*  add deps and scripts ([1c3dced0](https://github.com/lykmapipo/express-mquery/commit/1c3dced0151f044ff9649122eb0cc66f00ecd1ce))
*  implement on mquery on default export ([05c6cc05](https://github.com/lykmapipo/express-mquery/commit/05c6cc050fc9d0de3517cfb0674e118b993eff74))
*  implement middleware test to es6 ([2d5c0206](https://github.com/lykmapipo/express-mquery/commit/2d5c0206d8b4bac72d586ef838b72c5753f876f9))
*  implement examples to es6 ([f64131bb](https://github.com/lykmapipo/express-mquery/commit/f64131bbbdad1143c68aef4677e86aa59da40a77))
*  remove gruntfile and index ([1674d9cd](https://github.com/lykmapipo/express-mquery/commit/1674d9cd0649d6fcefa5ceb7c137b07eee779b84))
*  add nyc_output ([988e9037](https://github.com/lykmapipo/express-mquery/commit/988e9037223e533e97f87ba44123cf4a9ab23321))
*  ensure select field with data ([d3f9727f](https://github.com/lykmapipo/express-mquery/commit/d3f9727f643cdbd9686ad20488d9e07a37c0e83c))
*  migrate parser test to es6 ([b2dfdf10](https://github.com/lykmapipo/express-mquery/commit/b2dfdf10ecc7a8c911d9407bdfa62da00c1964a1))
*  migrate sort test to es6 ([b322e19f](https://github.com/lykmapipo/express-mquery/commit/b322e19f6d7931e40267ed430943869befea8302))
*  migrate select test to es6 ([21ebdaba](https://github.com/lykmapipo/express-mquery/commit/21ebdabad82dd6e01487f8b68f7b6c4838095882))
*  migrate populate test to es6 ([a98dd9b6](https://github.com/lykmapipo/express-mquery/commit/a98dd9b6cbf76ab81c3443c0adaf621fef49b6f8))
*  migrate paginate to es6 ([8d85fdb3](https://github.com/lykmapipo/express-mquery/commit/8d85fdb3dd135809b15c7f00c1808cfc8db7f4c1))

##### Refactors

* **internals:**
  *  ensure returned values ([c9772135](https://github.com/lykmapipo/express-mquery/commit/c97721357162647fc7dfc8a01026b50a2501a577))
  *  used named imports ([8df01d53](https://github.com/lykmapipo/express-mquery/commit/8df01d53cb500b515fabcdbba9e08fbe3096c9e7))
* **es6:**  improve package definition ([030d8766](https://github.com/lykmapipo/express-mquery/commit/030d87661c9df90624e4119e5036a8837cf6d4eb))
*  migrate to es6 ([f2121bf2](https://github.com/lykmapipo/express-mquery/commit/f2121bf2da4804a2ab7edc6e5d5e7fb2798f48d9))
*  setup rollup configuration file ([53a853ba](https://github.com/lykmapipo/express-mquery/commit/53a853ba7349cf394e291a47e304b40deb9206aa))
*  setup babel configuration ([8dc269a8](https://github.com/lykmapipo/express-mquery/commit/8dc269a8c91d16efaf56f34e4e59da7d56169637))

##### Code Style Changes

* **internals:**  improve jsdocs ([bf955bc3](https://github.com/lykmapipo/express-mquery/commit/bf955bc36714b819db8aa22eac1e31a0775b0fd0))

##### Tests

*  migrate headers to es6 ([61cc33a8](https://github.com/lykmapipo/express-mquery/commit/61cc33a860b64f468d5282071ae316066920353c))
*  migrate filter to es6 ([c9288f6c](https://github.com/lykmapipo/express-mquery/commit/c9288f6c72d23ed25e9cab928d8724305a2fd52b))
*  setup eslint configuration on unit test ([0d456962](https://github.com/lykmapipo/express-mquery/commit/0d456962eb3aec107417bf025bbb0e1118c79d87))

#### 1.2.14 (2021-02-16)

##### Chores

* **deps:**  force latest version & audit fix ([0251ae82](https://github.com/lykmapipo/express-mquery/commit/0251ae82cd4da54342aa44653b30c3a2bcaeae69))

#### 1.2.13 (2020-12-23)

##### Chores

* **deps:**

lib/index.js
'use strict';

const lodash = require('lodash');
const async = require('async');
const autoParse = require('auto-parse');

/**
 * @name filter
 * @function filter
 * @description parse filters from http query object into valid mongoose
 * query conditions
 * @param {object} options valid query params options to parse for
 * sorting conditions
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid mongoose(mongodb) query conditions(or criteria)
 * @see {@link http://jsonapi.org/format/#fetching-filtering}
 * @see {@link http://mongoosejs.com/docs/api.html#where_where}
 * @see {@link https://docs.mongodb.com/manual/reference/operator/query/}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.1.0
 * @public
 * @static
 * @example
 *
 * GET /customers?query={"name":"Bob"}
 * GET /customers?filter[name]=Bob
 * GET /customers?filter[name]={"$regex":"/Bo$/"}
 * GET /customers?filter[name][$regex]="/Bo$/"
 * GET /customers?filter[age]={"$gt":12}
 * GET /customers?filter[age][$gt]=12
 */
const filter = (options, done) => {
  // try parsing filters
  try {
    // ensure query
    const query = lodash.merge({}, autoParse(options));

    // obtain filters field
    const filterFields = query.filter || query.filters || query.query;
    const { q } = query;

    // check if is search query
    const filters = lodash.omitBy(
      lodash.merge({}, filterFields, { q }),
      (val) => typeof val !== 'boolean' && !val
    );

    // TODO ignore ['$jsonSchema', '$where']

    return done(null, filters);
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name headers
 * @function headers
 * @description parse headers to obtain request pre-conditions
 * @param {object} options valid http headers to parse for pre-conditions
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid parsed pre-conditions headers
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since}
 * @author lally elias <lallyelias87@mail.com>
 * @since 1.0.1
 * @version 0.1.0
 * @public
 * @example
 *
 * GET /customers?header['if-modified-since']=Mon Apr 16 2018 11:28:06 GMT+0300 (EAT)
 *
 * or
 *
 * curl -i -H 'If-Modified-Since: Wed, 12 Nov 2014 15:44:46 GMT' http://localhost:3000/invoices
 */
const headers = (options, done) => {
  // try parsing headers
  try {
    // ensure query
    const query = autoParse(options);
    let payload = lodash.merge({}, query.header, query.headers);

    // normalize headers

    // ... pick allowed headers
    const ALLOWED_HEADERS = ['if-modified-since'];
    payload = lodash.pick(payload, ALLOWED_HEADERS);

    // ... camelize header names
    payload = lodash.mapKeys(payload, (value, key) => lodash.camelCase(key));

    // ... normalize date headers
    const DATE_HEADERS = ['ifModifiedSince'];
    payload = lodash.mapValues(payload, (value, key) =>
      lodash.includes(DATE_HEADERS, key) ? new Date(value) : value
    );

    return done(null, payload);
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name paginate
 * @function paginate
 * @description parse paginations(i.e limit, offset, skip, page etc) from
 * http query object into valid mongoose pagination conditions
 * @param {object} options valid query params options to parse for sorting
 * conditions
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid mongoose(mongodb) pagination query
 * conditions(or criteria)
 * @see {@link http://jsonapi.org/format/#fetching-pagination}
 * @see {@link http://mongoosejs.com/docs/api.html#query_Query-skip}
 * @see {@link http://mongoosejs.com/docs/api.html#query_Query-limit}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.1.0
 * @public
 * @example
 *
 * GET /customers?skip=10&limit=10
 * GET /customers?limit=10
 * GET /customers?page=1
 * GET /customers?page[number]=1
 * GET /customers?page[number]=1&page[offset]=10
 * GET /customers?page[number]=1&page[limit]=10
 * GET /customers?page[number]=1&page[size]=10
 */
const paginate = (options, done) => {
  // try parsing paginations
  try {
    // ensure query
    let query = lodash.merge({ maxLimit: 50 }, autoParse(options));

    // normalize json-api page params
    if (query.page && lodash.isPlainObject(query.page)) {
      query = lodash.merge({}, query, query.page);
      query.page = query.number;
    }

    // initialize pagination
    let paginations = { limit: 10, skip: 0, page: 1 };

    // parse limit
    // see https://apigee.com/about/tags/restful-api-design
    let limit =
      query.limit ||
      query.max ||
      query.size ||
      query.rpp ||
      query.count ||
      query.perPage ||
      lodash.get(query, 'per_page') ||
      query.cursor ||
      10;
    limit = limit > query.maxLimit ? query.maxLimit : limit;
    limit = limit < 0 ? 10 : limit;
    paginations = lodash.merge({}, paginations, { limit });

    // parse page
    // see https://apigee.com/about/tags/restful-api-design
    let page = query.page || query.number || 1;
    page = page < 0 ? 1 : page;
    paginations = lodash.merge({}, paginations, { page });

    // parse skip
    // see https://apigee.com/about/tags/restful-api-design
    const hasSkip =
      lodash.has(query, 'skip') || lodash.has(query, 'offset') || lodash.has(query, 'start');
    let skip = query.skip || query.offset || query.start || 0;
    skip = skip < 0 ? 0 : skip;
    // correct skip base on page number
    skip = !hasSkip && page > 0 ? (page - 1) * limit : skip;
    paginations = lodash.merge({}, paginations, { skip });

    return done(null, paginations);
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name select
 * @function select
 * @description parse fields from http query object into valid mongoose
 * query select conditions
 * @param {object} options valid query params options to parse for select fields
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid mongoose(mongodb) query select conditions(or criteria)
 * @see {@link http://jsonapi.org/format/#fetching-sparse-fieldsets}
 * @see {@link http://mongoosejs.com/docs/api.html#query_Query-select}
 * @see {@link https://docs.mongodb.com/manual/tutorial/project-fields-from-query-results/}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.2.0
 * @public
 * @example
 *
 * GET /users?select=name
 * GET /users?select=name,email
 * GET /users?select=-name
 * GET /users?select=-name,-email
 * GET /users?select={"name":1}
 * GET /users?select={"name":1, "email":1}
 * GET /users?select={"name":0}
 * GET /users?select={"name":0, "email": 0}
 *
 * or
 *
 * GET /users?fields=name
 * GET /users?fields=name,email
 * GET /users?fields=-name
 * GET /users?fields=-name,-email
 *
 * or
 *
 * GET /users?select={"location.name":0, "location.address": 0}
 * GET /users?fields[location]=-name,-address
 */
const select = (options, done) => {
  // try parsing projections
  try {
    // ensure query
    const query = lodash.merge({}, autoParse(options));
    let projections = query.select || query.fields || query.projections;

    if (lodash.isString(projections)) {
      projections = [].concat(projections);
    }

    if (lodash.isPlainObject(projections)) {
      projections = lodash.map(projections, (proj, key) => {
        if (!lodash.isNaN(Number(key))) {
          return proj;
        }
        return {
          [key]: proj,
        };
      });
    }

    // parse comma separated fields into mongodb
    // awase projections
    const fieldify = (projection = '') => {
      // prepare projections
      let fields = lodash.compact(projection.split(','));
      fields = lodash.map(fields, lodash.trim);
      fields = lodash.uniq(fields);
      const accumulator = {};

      lodash.forEach(fields, (field) => {
        // if exclude e.g -name
        if (field[0] === '-') {
          accumulator[field.substring(1)] = 0;
        }

        // if include i.e name
        else {
          accumulator[field] = 1;
        }
      });

      return accumulator;
    };

    // normalize projections to object based
    projections = lodash.reduce(
      projections,
      (accumulator, projection) => {
        // handle string projection i.e select=name,email
        if (projection && lodash.isString(projection)) {
          const fields = fieldify(projection);
          // eslint-disable-next-line no-param-reassign
          accumulator = lodash.merge({}, accumulator, fields);
        }

        if (lodash.isPlainObject(projection)) {
          lodash.forEach(projection, (value, key) => {
            if (!lodash.isNaN(Number(value))) {
              // eslint-disable-next-line no-param-reassign
              accumulator = lodash.merge({}, accumulator, {
                [key]: value,
              });
            } else {
              const fields = fieldify(value);
              // eslint-disable-next-line no-param-reassign
              accumulator = lodash.merge({}, accumulator, {
                [key]: fields,
              });
            }
          });
        }

        return accumulator;
      },
      {}
    );

    return done(null, projections);
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name populate
 * @function populate
 * @description parse includes(or population) from http query object
 * into valid mongoose query population conditions
 * @param {object} options valid query params options to parse for
 * population or includes
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid mongoose(mongodb) query population
 * conditions(or criteria)
 * @see {@link http://jsonapi.org/format/#fetching-includes}
 * @see {@link http://mongoosejs.com/docs/api.html#query_Query-populate}
 * @see {@link http://mongoosejs.com/docs/populate.html#deep-populate}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.1.0
 * @public
 * @example
 *
 * /invoice?populate=customer
 * /invoice?populate=customer,items
 * /invoice?populate=customer.name,items.name,items.price
 * /invoice?populate={"path":"customer", "select":"name,price" }
 * /invoice?populate={"path":"customer", "select":{"name":1, "price":1} }
 * /invoice?populate=[{"path":"customer"}, {"path":"items"}]
 * /invoice?populate=[{"path":"customer", "select":"name"}, {"path":"items", "select":{"name": 1, "price": 1}}]
 *
 * or
 *
 * /invoice?includes=customer
 * /invoice?includes=customer,items
 * /invoice?includes=customer.name,items.name,items.price
 * /invoice?includes={"path":"customer", "select":"name,price" }
 * /invoice?includes={"path":"customer", "select":{"name":1, "price":1} }
 * /invoice?includes=[{"path":"customer"}, {"path":"items"}]
 * /invoice?includes=[{"path":"customer", "select":"name"}, {"path":"items", "select":{"name": 1, "price": 1}}]
 *
 * or
 *
 * /invoice?includes[customer]=name,number&includes[items]=name,price
 * /invoice?includes=customer,items&fields[customer]=name,number&fields[items]=name,price
 */
const populate = (options, done) => {
  // try parsing population
  // TODO support population on related object fields selection
  // TODO support nested population
  try {
    // ensure query
    const query = lodash.merge({}, autoParse(options));
    let populations = query.populate || query.include || query.includes;

    // handle string populations(or includes)
    // i.e includes=customer,items or
    // includes=customer.name,items.name,items.price
    if (populations && lodash.isString(populations)) {
      // ensure unique populations
      populations = [].concat(populations.split(','));
      populations = lodash.compact(populations);
      populations = lodash.map(populations, (population) => population.trim());
      populations = lodash.uniq(populations);

      // normalize populations to object based
      // and
      // map to array of populations
      populations = lodash.map(populations, (population) =>
        lodash.merge({}, { path: population })
      );
    }

    // handle object population
    // i.e {"path":"customer", "select":"name,price" }
    // or {"path":"customer", "select":{"name":1, "price":1} }
    // or
    // handle array populations
    // i.e [{"path":"customer", "select":"name,price" }]
    // or [{"path":"customer", "select":{"name":1, "price":1} }]

    // ensure array
    populations = lodash.compact([].concat(populations));

    // build populations
    return async.waterfall(
      [
        // parse field selections
        (next) => {
          select(query, next);
        },

        // normalize populations
        (selections, next) => {
          populations = lodash.map(populations, (population) => (cb) => {
            // obtain population selected fields
            const fields = selections ? selections[population.path] : undefined;
            if (fields) {
              // eslint-disable-next-line no-param-reassign
              population = lodash.merge({}, population, { select: fields });
            }

            // parse path specific select
            if (population.select) {
              select(population, (error, data) => {
                cb(error, lodash.merge({}, population, { select: data }));
              });
            }

            // continue
            else {
              cb(null, population);
            }
          });

          async.parallel(populations, next);
        },
      ],
      done
    );
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name sort
 * @function sort
 * @description parse sorts from http query object into valid mongoose sorting
 * conditions
 * @param {object} options valid query params options to parse for sorting
 * conditions
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} valid mongoose(mongodb) sorting conditions(or criteria)
 * @see {@link http://jsonapi.org/format/#fetching-sorting}
 * @see {@link http://mongoosejs.com/docs/api.html#query_Query-sort}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.1.0
 * @public
 * @example
 * GET /users?sort=name
 * GET /users?sort=name,email
 * GET /users?sort=-name
 * GET /users?sort=-name,-email
 * GET /users?sort={"name":1}
 * GET /users?sort={"name":1, "email":1}
 * GET /users?sort={"name":0}
 * GET /users?sort={"name":0, "email": 0}
 * GET /users?sort={"name":"asc", "email": "desc"}
 * GET /users?sort={"name":"ascending", "email": "descending"}
 */
const sort = (options, done) => {
  // try parsing sorts
  try {
    // ensure query
    const query = lodash.merge({}, autoParse(options));
    let sorts = query.order || query.sort;

    // handle string sort i.e sort=name,email
    if (sorts && lodash.isString(sorts)) {
      // ensure unique sorts
      sorts = lodash.compact(sorts.split(','));
      sorts = lodash.uniq(sorts);

      // normalize sorts to object based
      sorts = lodash.reduce(
        sorts,
        (accumulator, data) => {
          data.trim();

          // if exclude i.e sort=-name
          if (data[0] === '-') {
            accumulator[data.substring(1)] = -1;
          }

          // if include i.e sort=name
          else {
            accumulator[data] = 1;
          }

          return accumulator;
        },
        {}
      );
    }

    return done(null, sorts);
  } catch (error) {
    // catch all errors
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name parse
 * @function parse
 * @description parse specified JSON string into Javascript value or object
 * @param {string} string valid json string
 * @param {Function} done a callback to invoke on success or failure
 * @returns {object} constructed JavaScript value or object from JSON key value
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 0.1.0
 * @public
 */
const parse = (string, done) => {
  // TODO navigate paths and change them to valid mongodb query

  // try to parse json string
  try {
    const json = autoParse(string);

    return async.parallel(
      {
        filter: (next) => filter(json, next),
        headers: (next) => headers(json, next),
        paginate: (next) => paginate(json, next),
        populate: (next) => populate(json, next),
        select: (next) => select(json, next),
        sort: (next) => sort(json, next),
      },
      (error, payload) => {
        if (payload) {
          // remove unused
          // eslint-disable-next-line no-param-reassign
          delete payload.maxLimit;

          // remove populations in selects
          if (payload.populate && payload.select) {
            const populations = lodash.map(payload.populate, 'path');
            // eslint-disable-next-line no-param-reassign
            payload.select = lodash.omit(payload.select, populations);
          }

          // compact
          // eslint-disable-next-line no-param-reassign
          payload = lodash.omitBy(payload, lodash.isEmpty);
        }

        done(error, payload);
      }
    );
  } catch (error) {
    error.message = error.message || 'Bad Request';
    error.status = 400;
    return done(error);
  }
};

/**
 * @name mquery
 * @function mquery
 * @description parse http query parameters into valid mongoose query options
 * @param  {object} [optns] valid mquery options
 * @param  {object} [optns.limit] default limit
 * @param  {object} [optns.maxLimit] default max limit
 * @returns {Function[]} valid express middleware stack
 * @example
 *
 * import express from 'express';
 * import mquery from 'express-mquery';
 *
 *
 * const app = express();
 * app.use(mquery({limit: 10, maxLimit: 50}));
 */
const mquery = (optns) => {
  // normalize options
  const options = lodash.merge(
    {},
    {
      limit: 10,
      maxLimit: 50,
    },
    optns
  );

  // pack & return middlewares stack
  return [
    (request, response, next) => {
      const query = lodash.merge(
        {},
        options,
        { headers: request.headers },
        request.query
      );
      parse(query, (error, data) => {
        if (data && !lodash.isEmpty(data)) {
          request.mquery = data;
        }
        next(error, data);
      });
    },
  ];
};

/**
 * @module expess-mquery
 * @name expess-mquery
 * @description Expose mongoose query API through HTTP request
 * @see {@link https://docs.mongodb.com/manual/reference/operator/query/}
 * @see {@link http://mongoosejs.com/docs/api.html#Query}
 * @see {@link http://jsonapi.org/recommendations/}
 * @see {@link http://mgmco.github.io/api-query-spec/}
 * @author lally elias <lallyelias87@mail.com>
 * @since 0.2.2
 * @version 1.3.0
 * @public
 */

module.exports = mquery;
